libconnman-qt/networkservice.cpp

```diff
@@ -792,42 +788,45 @@ void NetworkService::Private::setLastConnectError(const QString &error)
     }
 }
 
-void NetworkService::Private::updateConnecting(const QString &state)
-{
-    bool connecting = m_connectWatcher || ConnmanState::connecting(state);
-    if (m_connecting != connecting) {
-        m_connecting = connecting;
-        queueSignal(SignalConnectingChanged);
+static NetworkService::ServiceState stateStringToEnum(const QString &state)
+{
+    if (state == QStringLiteral("idle")) {
+        return NetworkService::IdleState;
+    } else if (state == QStringLiteral("failure")) {
+        return NetworkService::FailureState;
+    } else if (state == QStringLiteral("association")) {
+        return NetworkService::AssociationState;
+    } else if (state == QStringLiteral("configuration")) {
+        return NetworkService::ConfigurationState;
+    } else if (state == QStringLiteral("ready")) {
+        return NetworkService::ReadyState;
+    } else if (state == QStringLiteral("disconnect")) {
+        return NetworkService::DisconnectState;
+    } else if (state == QStringLiteral("online")) {
+        return NetworkService::OnlineState;
     }
-}
 
-void NetworkService::Private::updateConnecting()
-{
-    updateConnecting(state());
+    return NetworkService::UnknownState;
 }
 
-void NetworkService::Private::updateConnected(const QString &state)
+void NetworkService::Private::updateState()
 {
-    bool connected = !m_connectWatcher && ConnmanState::connected(state);
-    if (m_connected != connected) {
-        m_connected = connected;
-        queueSignal(SignalConnectedChanged);
-    }
-}
+    QString currentState = state();
 
-void NetworkService::Private::updateConnected()
-{
-    updateConnected(state());
-}
+    bool wasConnecting = service()->connecting();
+    bool wasConnected = service()->connected();
 
-void NetworkService::Private::updateState()
-{
-    QString currentState(state());
-    if (m_state != currentState) {
-        m_state = currentState;
-        queueSignal(SignalStateChanged);
-        updateConnecting(currentState);
-        updateConnected(currentState);
+    m_serviceState = stateStringToEnum(currentState);
+
+    queueSignal(SignalStateChanged);
+    queueSignal(SignalServiceStateChanged);
+
+    if (service()->connecting() != wasConnecting) {
```

Comment:
I'm not following now, does the `connecting` and `connected` get changed in between this and the earlier assignment? The signals are queued and this is relying and checking if the signal has been even been processed out of the queue AND something changed on this service due to that? Just wondering about the async nature here in Qt that can it even happen and does the change be done then too late that this state is not updated - or will it be? Again this is a big change...

Reply:
It's not really that big of a change. What this is about is that both connecting and connected depend on the state, and thus if the state is changed we need to check if that affected the results of these.

Reply:
Also to note that the old code did the same checking, it was just having separate methods that were called from this one. What I did there is getting rid of generally redundant member variables for these states.

Reply:
> It's not really that big of a change. 

I meant this PR as total. But yeah, should be ok.